### PR TITLE
Fixes repair compact

### PR DESCRIPTION
**Pull request by ProgrammerDan**
Just going to leave this open, perhaps people could springboard off of it. There are likely errors so don't merge into master just yet. It builds fine however. The issues I'm targetting:

* Shaped and Shapeless recipes aren't working

* Repair factory can't be repaired (fixed: can be repaired now in this commit)

* Compactor had no Crates (fixedish: CivTest only recipes)

* Repair factory eats too much materials eventually (unconfirmed locally)

* Better logging of what FactoryMod is doing

* No message to users on repair factory successful create

* Null stack trace if no factory could be created on valid-block-config-but-not-right-contents (fixed with nice message to user)

What else am I missing? This stuff is bare minimum for out the door.

### CODE REVIEW

**Comment by ProgrammerDan** [comment]
Just a reminder: don't merge. Just leaving this here if someone who doesn't go to work wants to stab at it.

**Comment by rourke750** [comment]
I was gone during the past week but how did everything break?

**Comment by ProgrammerDan** [comment]
Well, IO Refactor tossed a lot of stuff into disarray. I think those bugs are mostly squashed (except for the recipes, still trying to figure that out), but they hid a few bugs within repair factory.

Also, confusion over XP rewrite led to pushing Teal's compactor through into master, it is largely untested and likely has a few hidden bugs (although less critical than getting repair out). 

So, I have a civtest-only recipe for Compactor in this pull, along with a few fixes to Repair Factory (it was using the sum of repair ingredients and reset ingredients, you need that switching logic in getInputs() and getOutputs() to prevent overlap) and the start of some fixes for recipes and the other stuff above. Logging as well, because frankly the logging in FactoryMod is subpar, wire-wrapping circuit boards is simpler than tracing errors here.

**Comment by rourke750** [comment]
Shouldn't the first step be to revert to before it was broken and then only
have to worry about the io not working.

**Comment by ProgrammerDan** [comment]
When which thing wasn't working? That's the issue, if I could isolate it simply like that I would have. 

Far as I can tell, IO Refactor didn't introduce many regressions and @WildWeazel has gotten most of those. The bugs I'm chasing where already present in Repair Factory. 

The only exception might be the crafting recipes, but it's acting more like underlying API changes (some stuff was deprecated and we were using the old methods)

**Comment by rourke750** [comment]
So the only issue I have with that statement is that I believe craft
recipes still work on civcraft which is running the same as civtest. Also
from when I had finished the repair factory some of these bugs I heard were
not present then.

**Comment by rourke750** [comment]
Either way you know more what's going on and I'll leave it to you to handle.

**Comment by ProgrammerDan** [comment]
I think Civcraft is pretty far back right now, could be wrong (pre-IO refactor, for instance). 

I'll doublecheck from before the IO refactor ref: repair factory. It is definitely possible that someone manually merged an old commit and blew something away.

**Comment by WildWeazel** [comment]
I think some of the regressions are because the new components had bugs in them that cause certain features to abort prematurely, eg some new factory creation was failing because the repair factory attempt threw a NPE.

I finally got my machine set up for full server testing again, so I can more quickly run through the rest of my bug fixes.

**Comment by ProgrammerDan** [comment]
In the same position @WildWeazel finally got the VM running with Namelayer and Citadel, so I was able to make some progress from 1am on. 

https://github.com/Civcraft/FactoryMod/blob/master/src/com/github/igotyou/FactoryMod/managers/AManager.java#L109

This line is why there is no message on create, Repair Factory is simply returning the superclass return and doesn't decorate it with a custom message.

**Comment by ProgrammerDan** [comment]
Double checked blame just to make sure I didn't miss a regression, the only changes to RepairFactory with respect to IO Refactor are the superclass/interface dependencies. Sorry :( these errors just missed detection. No worries, we'll knock 'em out.

**Comment by rourke750** [comment]
Did my recent changes with NameLayer help with building with maven?

**Comment by ProgrammerDan** [comment]
Your latest commits came after I gave up and used March 23rd's builds :(. I'll try again with latest tonight and let you know.

**Comment by rourke750** [comment]
Lol. I changed how it commits and moved all nms code out of NameLayer that
referenced old versions so that it could be built locally.

**Comment by ProgrammerDan** [comment]
Haha, yes, that should help immensely.

**Comment by ProgrammerDan** [comment]
@WildWeazel I'm working on this now too, lemme know if you've got fixes I should integrate.

**Comment by ttk2** [comment]
how is this coming?

**Comment by ProgrammerDan** [comment]
lol so... the repair factory "repairs" excess items in the chest, adding NBT data to them and making them "unique" compared to their unrepaired brothers ... and also unusable as repair materials. Fixing.

**Comment by WildWeazel** [comment]
I can't repair Compactors, is that expected? It doesn't seem to recognize emeralds in the chest.

**Comment by ProgrammerDan** [comment]
Not sure yet. Which recipe are you using, and which commit?

**Comment by WildWeazel** [comment]
On Civtest, 64D blocks to create then 16 emeralds to repair from 0

**Comment by ProgrammerDan** [comment]
Ah, ok, yeah. The civtest one is very broken. I think we've plumped those depths, I'm in break-fix mode now.

There's a few serious breaking bugs here. I'm furiously fixing them. 

Recipe disabling was not working -- is working correctly now and outputs sensible information to the logs -- had to rewrite

Create success for compactor and repair should work now -- not yet confirmed but code looks good

Repair undiscriminately writes NBT data on anything in the repair chest. If you have 'too much" repair material, it alters their NBT causing them to be useless for future use, presumably unless placed and re-obtained (if possible). It does cause stacking to break for those items -- In progress (big change) -- and there may be other issues here. First run of the factory took 128 coal ore and 8 emerald blocks, in spite of recipe only calling for 64 coal ore and 8 emerald blocks; that's a big issue.

**Comment by ProgrammerDan** [comment]
@rourke750 @ttk2 @WildWeazel this latest merge is good to build and put on civtest.

A summary of my changes:

1. Recipe disabling has improved matching, and works well, and doesn't disable the wrong recipes, hurray. **I'd like people to test these as described in the testing thread**

2. Unshaped recipes work (Emerald to Bottle, and back) but shaped recipes aren't working yet. Investigating. **I'd like people to verify this**

3. Repair factory's "reset" is working great now. I had to layer in specific allowances for repairable items; it's now a configuration section of the repair factory called "repairables" so if Minecraft adds new stuff we can easily add it. It matches *only* on material type, so if we need repairables at specific durabilities or other item data this will need improving. **I've tested it minimally locally, it looks good, but I need someone to test every repairable **

4. Repair factory's "reset" now correctly only consumes the inputs once. For some reason, the super class's consumer, and the subclass consumer were both called. This has been corrected. **Would like testing on this.**

5. Repair factory's "repair" is still wonky. For no damage it will consume 3 coal ore to repair ... 0%. For fully damaged, it will consume 6 coal ore. In both cases it updates 4x as many as the max repair is supposed to be. I'm still investigating, but not finding it yet.

6. Repair factory would have been creatable with "more than the construction costs" in the chest. It now uses the proper function to ensure that only the construction items are in the chest when creating the factory. **Please test**

7. Compactor factory had the same bug. It should have the same correction, I have not tested it. **Please test**

8. Some other general cleanup, tried to restrain myself to focus on critical changes.

**Comment by ttk2** [comment]
ok to test 

getting a jar and its going straight on Civtest, thanks dan.

**Comment by ttk2** [comment]
it seems I can't get a build of this without merging it, oh well, here goes nothing.

**Comment by rourke750** [comment]
:/

File: src/com/github/igotyou/FactoryMod/managers/CompactorManager.java [hunk 1]
```diff
@@ -22,25 +22,25 @@ public CompactorManager(FactoryModPlugin plugin) {
...
+            Chest chest = (Chest) inventoryBlock.getState();
+            Inventory chestInventory = chest.getInventory();
+            ItemList<NamedItemStack> inputs = compactorProperties.getConstructionMaterials();
+
+			if (inputs.exactlyIn(chestInventory)) {
 				Compactor production = new Compactor(factoryLocation, inventoryLocation, powerLocation, false, plugin.getCompactorProperties());
-				if (compactorProperties.getConstructionMaterials().removeFrom(production.getInventory()))
-				{
-					addFactory(production);
+				inputs.removeFrom(production.getInventory());
```
Comment (ProgrammerDan) [inline src/com/github/igotyou/FactoryMod/managers/CompactorManager.java]: Note to self: Now that super.addFactory() is a boolean function, removal of recipe ingredients should happen *after* that method returns true. E.g. move this down two lines. Same for Repair Factory.

File: src/com/github/igotyou/FactoryMod/Factorys/RepairFactory.java [hunk 2]
```diff
@@ -242,23 +239,38 @@ public double getProductionTime() {
...
+			if (stack == null) {
 				continue;
-			ItemStack s = CraftItemStack.asNMSCopy(stack);
-			s.setRepairCost(1);
-			getInventory().setItem(x, CraftItemStack.asBukkitCopy(s));
+			}
+			if (repairable(stack)) {
+				log.fine("Found repairable: " + stack.getType());
+				ItemStack s = CraftItemStack.asNMSCopy(stack);
+				s.setRepairCost(1);
```
Comment (ProgrammerDan) [inline src/com/github/igotyou/FactoryMod/Factorys/RepairFactory.java]: @ttk2 This balance discussion for Reset factory could be completely short-circuited if I changed this number to 25. Then, no matter the tool, once you cross that 3 anvil-repair threshold, the best you can achieve is a 27lvl repair on the reset factory. Then we could make the reset factory reset recipe basically free and stop dickering and get this online.

Thoughts?

Comment (ttk2) [inline src/com/github/igotyou/FactoryMod/Factorys/RepairFactory.java]: I don't want to make things cheaper if I can avoid it. If its not the best option we can just put online somthing that will piss people off. 

If its not relevant anyways go for it.

**Comment by WildWeazel** [comment]
For future reference I think the comment has to be only the trigger phrase

**Comment by WildWeazel** [comment]
ok to test

**Comment by WildWeazel** [comment]
or you know, it could be broken

**Comment by rourke750** [comment]
Been some bugs in the latest updates.  It should still build them though
and I believe it links by the pull request merge thing.

**Pull request merged**